WiinUPro/Controls/NintyControl.xaml.cs

```diff
@@ -627,6 +646,10 @@ private void btnAddRumble_Click(object sender, RoutedEventArgs e)
             }
 
             _rumbleSubscriptions = win.Result;
+            if (AppPrefs.Instance.autoAddXInputDevices)
+            {
+                MainWindow._instance.AddXInput(_rumbleSubscriptions);

```

Comment:
Instead of reaching into this class instance, pipe it through an event similar to how OnPrefsChange just below.

Reply:
How would one go about doing that? OnPrefsChange below is accessing a newly created window, whereas I want to access MainWindow. How would one access the original MainWindow? Would one somehow pass the function from MainWindow to NintyControl through DeviceStatus? That seems like a not so straightforward method, although I did get that to work by passing the function like so:
![image](https://user-images.githubusercontent.com/63229554/95448034-fbbf5f00-0962-11eb-9925-8f51d7077214.png)
![image](https://user-images.githubusercontent.com/63229554/95448817-337ad680-0964-11eb-93b9-4e00cf18c3a4.png)
When calling the function from NintyControl:
![image](https://user-images.githubusercontent.com/63229554/95448898-4a212d80-0964-11eb-8587-302da0755063.png)
I have zero experience with using events and actions, help would be appreciated. Is this solution similar to the one you had in mind?

Reply:
That's looking right. Just to give a full overview:

Declare this in NintyControl
`public event Action<bool[]> OnRumbleSubscriptionChange`

And call `OnRumbleSubscriptionChange?.Invoke(_rumbleSubscriptions)` anywhere you were calling into MainWindow._instance.AddXInput(_rumbleSubscriptions)

In DeviceStatus.xaml.cs declare
`public Action<DeviceStatus, bool[]> OnRumbleSubscriptionChange;`

Then add an event listener where a new NintyControl is being created in the constructor
`Ninty.OnRumbleSubscriptionChange += RumbleSettingsChanged;` (or whatever method name you think makes sense).

And in that method you would call 
`OnRumbleSubscriptionChange?.Invoke(this, rumbleSubscriptions)`

Finally in MainWindow.xaml.cs around line 164 define the callback for OnRumbleSubscriptionChange which will be your AddXInput function logic.

This may look very quite chainy and seems unnecessary when you could simply just reach into the MainWindow directly, but look at it like the MainWindow is now has a method dedicated for when this action happens on a device instead of just freely letting anything call into it. MainWindow is dictating what to do with this information and the infrastructure will also gracefully handle supporting this feature for JoyControls or other potential new control schemes to use rumble subscription in the future.

I believe you just commit it to your fork, and potentially update the pull request? I'm not sure as I haven't done so in GitHub myself.


On a side note, I see I have some refactoring and cleaning up I should do as well as adding comments to make this code more maintainable.